consensus/src/round_manager_test.rs

```diff
@@ -835,3 +835,55 @@ fn sync_on_partial_newer_sync_info() {
         let _ = node.next_sync_info().await;
     });
 }
+
+#[test]
+fn safety_rules_crash() {
+    let mut runtime = consensus_runtime();
+    let mut playground = NetworkPlayground::new(runtime.handle().clone());
+    let mut nodes = NodeSetup::create_nodes(&mut playground, runtime.handle().clone(), 1);
+    let mut node = nodes.pop().unwrap();
+    runtime.spawn(playground.start());
+
+    fn reset_safety_rules(node: &mut NodeSetup) {
+        let safety_storage = PersistentSafetyStorage::initialize(
+            Storage::from(libra_secure_storage::InMemoryStorage::new()),
+            node.signer.author(),
+            node.signer.private_key().clone(),
+            Ed25519PrivateKey::generate_for_testing(),
+            node.round_manager.consensus_state().waypoint(),
+        );
+
+        node.safety_rules_manager = SafetyRulesManager::new_local(safety_storage, false);
+        let safety_rules =
+            MetricsSafetyRules::new(node.safety_rules_manager.client(), node.storage.clone());
+        node.round_manager.set_safety_rules(safety_rules);
+    };
+
+    timed_block_on(&mut runtime, async {
+        for _ in 0..2 {
+            let proposal_msg = node.next_proposal().await;
+
+            reset_safety_rules(&mut node);
+            // construct_and_sign_vote
+            node.round_manager
+                .process_proposal_msg(proposal_msg)
+                .await
+                .unwrap();
+
+            let vote_msg = node.next_vote().await;
+
+            // sign_timeout
+            reset_safety_rules(&mut node);
+            let round = vote_msg.vote().vote_data().proposed().round();
+            node.round_manager
+                .process_local_timeout(round)
+                .await
+                .unwrap();
+            let vote_msg = node.next_vote().await;
+
+            // sign proposal
+            reset_safety_rules(&mut node);
+            node.round_manager.process_vote_msg(vote_msg).await.unwrap();
```

Comment:
does process_vote_msg invoke safety rules?

Reply:
since there's only a single node, process_vote -> qc accumulation -> proposal generation -> sign_proposal

Reply:
can we unwrap next_proposal() to make it more explicit?

Reply:
that just reads it from the channel, but happy to do so :)